### PR TITLE
Fix-ish hyperopt leak

### PR DESCRIPTION
**Pull request by scarlehoff**
I haven't been able to actually fix the memory leak, so I took a different route. Instead of fixing the leak, now each hyperopt trial is run as forked process that dies once the trial finishes.

With this we can run hyperopt for as many trials we want without hitting OOM. I'm bypassing it when running hyperopt in parallel because in that case one can just leave the database running and run many small jobs and then it would be better not to have the overhead of these changes.

Not very beautiful but functional*. I've left also my previous attempts (in `internal_state.py`) at cleaning up tensorflow which were not enough but they did clean up things that were dangling so I think it might be worth having them.

(tested by submitting many jobs at BSC and none of them failed due to an OOM, which was the case before! I'll add an actual benchmark)

### CODE REVIEW

**Comment by Radonirinaunimi** [comment]
Thanks @scarlehoff for this. I will have a look ASAP before the end of this week.

**Comment by scarlehoff** [comment]
Thanks! Turns out this is going to be important because we need to redo the hyperopt with MHOU (which certainly requires be very light on memory :__ )

**Comment by scarlehoff** [comment]
The promised benchmark, running in `montblanc` in Milan, with the memory peaks:

A first run small run (so take the times with a grain of salt, since due to the computer and the size of the runcard, the overhead is irrelevant, the memory was certainly the cause of the difference)
```
master: 22 GB, 2.5 hours

this branch:  16 GB, 30 minutes
```

A second run, with a greater number of datasets and replicas

```
master: nothing, OOM, i got a peak of 28 GB before the OOM which happened after 1 hours

this branch: 24 GB, it finished after 6 hours
```

File: n3fit/src/n3fit/backends/keras_backend/internal_state.py [hunk 1]
```diff
@@ -18,8 +18,16 @@
 from keras import backend as K
 import numpy as np
 
+from validphys.convolution import central_predictions, predictions
```
Comment (Radonirinaunimi) [inline n3fit/src/n3fit/backends/keras_backend/internal_state.py]: It seems to me that this now violates the entire purpose of `internal_state.py`. It is supposed to be a generic backend but now it imports `validphys.convolution` and nukes its process-wide cache on every trial.

For the sake of maintainability, I think it would be best to pass `central_predictions.cache_clear` and `predictions.cache_clear` as arguments to `clear_backend_state` and hence avoid the module import.

File: n3fit/src/n3fit/backends/keras_backend/internal_state.py [hunk 2]
```diff
@@ -62,6 +71,14 @@ def set_threading(threads, cores, double_precision=False):
                 "Could not set tensorflow parallelism settings from n3fit, maybe tensorflow is already initialized by a third program"
             )
 
+    def _internal_backend_clear():
+        """TensorFlow saves the gradient of the custom models we create as custom gradients.
+        These are not followed by keras and are thus not cleared during the clear_backend call."""
+        registry = ops.gradient_registry._registry
```
Comment (Radonirinaunimi) [inline n3fit/src/n3fit/backends/keras_backend/internal_state.py]: I am always hesitant about the usage of these private API without safeguards. Can we at least enclose it in a try/except + warnings? That'd be better than a hard failure when this changes in newer versions.

File: n3fit/src/n3fit/hyper_optimization/hyper_scan.py [hunk 3]
```diff
@@ -45,9 +48,55 @@
...
+        finally:
+            sender.close()
+
+    process = context.Process(target=_run_objective)
+    process.start()
+    # The parent only receives, close its sender
+    sender.close()
+    try:
+        # Wait for the child to finish running
+        success, ret = receiver.recv()
```
Comment (Radonirinaunimi) [inline n3fit/src/n3fit/hyper_optimization/hyper_scan.py:83]: In principle, for some god knows what reasons, a child process could deadlock, causing the job to hang (with no way of telling if it's stuck or slow) (?). Should we perhaps add a timeout to avoid this?

**Review by Radonirinaunimi — COMMENTED** [review]
This looks generally fine to me. I only have very minor comments/suggestions below and they are mainly pedantic aspects. What I think are really missing though are unit tests, especially for `_run_trial_in_subprocess`.

We could for example have simple tests that:
- the actual forks really have different PID than the parent
- in case of no forks/macos it falls back directly to the current process
- the timeout is enforced
- ...

File: n3fit/src/n3fit/backends/keras_backend/internal_state.py (hunk 2 above)
Comment (scarlehoff) [reply n3fit/src/n3fit/backends/keras_backend/internal_state.py]: I'll do the `try-except` (although fwiw it could even be removed, its effect is tiny in comparison with the separate process)

File: n3fit/src/n3fit/hyper_optimization/hyper_scan.py (hunk 3 above)
Comment (scarlehoff) [reply n3fit/src/n3fit/hyper_optimization/hyper_scan.py:83]: You are right, but at the same time we don't know beforehand we don't know how long a particular architecture should take.
I _think_ this is a problem that will be solved by the cluster admin's email saying "your job has been blocking a GPU at 0% usage for the last 24 hours" :P

(more seriously, I wouldn't know how to solve this problem effectively)

File: n3fit/src/n3fit/backends/keras_backend/internal_state.py (hunk 1 above)
Comment (scarlehoff) [reply n3fit/src/n3fit/backends/keras_backend/internal_state.py]: Done (but since it is only these two, I've opted for keeping themin the model trainer hyperoptimization)

Comment (scarlehoff) [reply n3fit/src/n3fit/backends/keras_backend/internal_state.py]: Done (but since it is only these two, I've opted for keeping themin the model trainer hyperoptimization)

File: n3fit/src/n3fit/hyper_optimization/hyper_scan.py (hunk 3 above)
Comment (Radonirinaunimi) [reply n3fit/src/n3fit/hyper_optimization/hyper_scan.py:83]: That's true. In most cases, it'd be even by the scheduler when reaching the maximum time. As I mentioned, this is really a very minor thing so we could disregard.

My thinking was that setting a "reasonable" threshold would be a slightly better safeguard in that it will free up resources for other trials/processes to run (instead of just blocking it). And by reasonable I simply mean a number that we can tolerate (ie do we even expect or tolerate a run that exceeds 20 hrs?).

Comment (scarlehoff) [reply n3fit/src/n3fit/hyper_optimization/hyper_scan.py:83]: We might want that if we are allowing for an extremely long number of epochs.

Comment (scarlehoff) [reply n3fit/src/n3fit/hyper_optimization/hyper_scan.py:83]: We might want that if we are allowing for an extremely long number of epochs.

**Review by Radonirinaunimi — APPROVED** [review]
